modules/programs/fish.nix

```diff
@@ -45,109 +21,163 @@ in
-          characters.
+          Whether fish should use completion files provided by other packages.
+        '';
+      };
+      
+      vendor.functions.enable = mkOption {
+        type = types.bool;
+        default = true;
+        description = ''
+          Whether fish should autoload fish functions provided by other packages.
         '';
-        apply = mapAttrs (n: v: if isList v then concatStringsSep ":" v else v);
       };
 
       shellAliases = mkOption {
-        type = types.attrs;
-        default = cfge.shellAliases;
+        default = config.environment.shellAliases;
         description = ''
           Set of aliases for fish shell. See <option>environment.shellAliases</option>
           for an option format description.
         '';
+        type = types.attrs;
       };
 
       shellInit = mkOption {
-        type = types.lines;
         default = "";
         description = ''
-          Shell Script code called during fish shell initialisation.
+          Shell script code called during fish shell initialisation.
         '';
+        type = types.lines;
       };
 
       loginShellInit = mkOption {
-        type = types.lines;
         default = "";
         description = ''
           Shell script code called during fish login shell initialisation.
         '';
+        type = types.lines;
       };
 
       interactiveShellInit = mkOption {
-        type = types.lines;
         default = "";
         description = ''
           Shell script code called during interactive fish shell initialisation.
         '';
+        type = types.lines;
       };
 
       promptInit = mkOption {
-        type = types.lines;
         default = "";
         description = ''
           Shell script code used to initialise fish prompt.
         '';
+        type = types.lines;
       };
 
     };
+
   };
 
   config = mkIf cfg.enable {
 
-    environment.systemPackages = [ pkgs.fish ];
-
-    environment.pathsToLink = [ "/share/fish" ];
-
-    environment.loginShell = "${shell}/bin/fish -l";
-    environment.variables.SHELL = "${shell}/bin/fish";
+    environment.etc."fish/foreign-env/shellInit".text = cfge.shellInit;
+    environment.etc."fish/foreign-env/loginShellInit".text = cfge.loginShellInit;
+    environment.etc."fish/foreign-env/interactiveShellInit".text = cfge.interactiveShellInit;
+    environment.etc."fish/foreign-env/extraInit".text= cfge.extraInit;
+
+    environment.etc."fish/nixos-env-preinit.fish".text = ''
+      # This happens before $__fish_datadir/config.fish sets fish_function_path, so it is currently
+      # unset. We set it and then completely erase it, leaving its configuration to $__fish_datadir/config.fish
+      set fish_function_path ${pkgs.fish-foreign-env}/share/fish-foreign-env/functions $__fish_datadir/functions
+     
+      # For "reasons not remembered," setEnvironment does not set PATH, so do it here
+      set PATH ${replaceStrings [":"] [" "] config.environment.systemPath} 
```

Comment:
I think it was something related to ordering, but I guess it's actually useful in this case. How is the separator handled on nixos?

Reply:
Since setting the PATH happens inside `fenv`'s temporary bash environment, the foreign-env plugin for _Oh My Fish!_ handles [converting the separator for us](https://github.com/oh-my-fish/plugin-foreign-env/blob/master/functions/fenv.apply.fish#L31). This applies just as well to nix-darwin since we're also using fenv. :-)

Reply:
Ah, interesting. I might change it to just set `environment.variables.PATH` at some point then.